actions: run: Check script exists during recipe verification

obbardc opened the pull request:
If a script doesn't exist or the permissions are wrong, the
recipe exits during execution. Let's check early if the
script exists on the host filesystem and that the file has
executable permission bit set.

sjoerdsimons commented on actions/run_action.go:
```diff
@@ -68,6 +70,22 @@ func (run *RunAction) Verify(context *debos.DebosContext) error {
+
+	if run.Script != "" {
+		argv := strings.SplitN(run.Script, " ", 2)
+		script := debos.CleanPathAt(argv[0], context.RecipeDir)
```
in PreMachine we roughly have the same code to determine the path of the script ; Probably worth havign a single utility function for it

obbardc replied:
in a separate commit i reworked that so we do that determination only once and earlier which improves readability to me

sjoerdsimons commented on actions/run_action.go:
```diff
@@ -68,6 +70,22 @@ func (run *RunAction) Verify(context *debos.DebosContext) error {
+		}
+
+		/* check the script is executable */
+		if stat.IsDir() || stat.Mode()&0111 == 0 {
```
what if the script is a symlink?

obbardc replied:
Split the executable/dir check into two

it also works; clarified that test in the commit message

sjoerdsimons commented:
fwiw this would be a great change to start adding test for

obbardc committed:
actions: run: Check script exists during recipe verification

If a script doesn't exist or the permissions are wrong, the recipe exits
during the recipe execution. Let's do some early checks in the Verify
stage to ensure the script exists on the host filesystem and a basic check
that the file has executable permission bit set. This check also follows
symlinks.

Signed-off-by: Christopher Obbard <chris.obbard@oss.qualcomm.com>

obbardc committed:
actions: run: Check script exists during recipe verification

If a script doesn't exist or the permissions are wrong, the recipe exits
during the recipe execution. Let's do some early checks in the Verify
stage to ensure the script exists on the host filesystem and a basic check
that the file has executable permission bit set. This check also follows
symlinks.

Signed-off-by: Christopher Obbard <christopher.obbard@linaro.org>

sjoerdsimons commented on actions/run_action.go:
```diff
@@ -120,7 +120,7 @@ func (run *RunAction) doRun(context debos.DebosContext) error {
 		cmdline = []string{run.scriptPath}
 		cmdline = append(cmdline, run.scriptArgs...)
-		label = path.Base(run.Script)
+		label = path.Base(run.scriptPath)
```
but then you don't see the arguments to the script anymore, which is a loss.. Probably this doesn't belong in this PR (it doesn't help with your stated target i think) but preventing "pollution" we could do either by:
* unify truncating the length like the Command handling does
* printing what gets run *first* and then either not or only providing a minimal label on each line after

obbardc replied:
dropped the commit

sjoerdsimons commented on actions/run_action.go:
```diff
@@ -82,6 +83,21 @@ func (run *RunAction) Verify(context *debos.DebosContext) error {
+			return err
+		}
+
+		if stat.IsDir() {
```
Rather then checking if it's a Dir this should proably check if it's a regular file

sjoerdsimons commented on actions/run_action.go:
```diff
@@ -82,6 +83,21 @@ func (run *RunAction) Verify(context *debos.DebosContext) error {
+		}
+
+		/* Check the script is executable */
+		if stat.Mode()&0111 == 0 {
```
executable and readable ideally; though i don't think golang gives you nice helpers for that unfortunately :/

sjoerdsimons commented on .github/workflows/ci.yaml:
```diff
@@ -136,6 +136,8 @@ jobs:
           - backend: kvm
             test: { name: "apertis", case: "apertis" }
+          - backend: kvm
+            test: { name: "run", case: "run" }
```
kvm actually has the highest latency atm because of the runner setup; and really this should probably run on all types.. That said could you merge it with an existing tests so we don't need a new CI job for it ? 

e.g. maybe merge it with the recipes test and the overlay test :)

obbardc committed:
tests: exit_code: Add various test cases for run action

Add new test cases to verify that run actions with scripts actually
call a script which exists on the filesystem as well as some other
interesting cases, like missing scripts, symlinked scripts and
scripts which aren't executable.

Signed-off-by: Christopher Obbard <christopher.obbard@linaro.org>

obbardc committed:
tests: exit_code: Add various test cases for run action

Add new test cases to verify that run actions with scripts actually
call a script which exists on the filesystem as well as some other
interesting cases, like missing scripts, symlinked scripts and
scripts which aren't executable.

Signed-off-by: Christopher Obbard <chris.obbard@oss.qualcomm.com>

obbardc committed:
actions: run: Fix import order

Signed-off-by: Christopher Obbard <chris.obbard@oss.qualcomm.com>

obbardc committed:
actions: run: Fix import order

Signed-off-by: Christopher Obbard <christopher.obbard@linaro.org>

obbardc commented:
@sjoerdsimons manually rebased on top of main; should be OK to go now.

obbardc commented:
@tobsan can you review this please? I've rebased & reworked onto your recent changes. Maybe I missed something obvious.

obbardc commented on actions/run_action.go:
```diff
@@ -104,18 +144,17 @@ func (run *RunAction) doRun(context debos.Context) error {
+
+		cmdline = []string{scriptPath}
+		cmdline = append(cmdline, scriptArgs...)
+		label = path.Base(run.Script)
```
good catch, thanks!

obbardc commented on actions/run_action.go:
```diff
@@ -64,14 +66,54 @@ type RunAction struct {
 	}
+
+	if run.Script != "" {
+		scriptPath, _ := run.extractScriptArgs(run.Script, context.RecipeDir)
```
for now I'd rather leave it in this function, but good point.

obbardc commented on actions/run_action.go:
```diff
@@ -131,18 +168,17 @@ func (run *RunAction) doRun(context debos.Context) error {
 	if run.Script != "" {
-		script := strings.SplitN(run.Script, " ", 2)
-		script[0] = debos.CleanPathAt(script[0], context.RecipeDir)
+		scriptPath, scriptArgs := run.extractScriptArgs(run.Script, context.RecipeDir)
```
need to make sure the latest rebase includes current logic in the run action file...
